Initial commit for validator

jvanstraten opened the pull request:
This PR is only for checking CI

jvanstraten committed:
Initial commit for validator

mbrobbel commented on .github/workflows/pr.yml:
```diff
@@ -64,3 +64,59 @@ jobs:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+      - run: rustup component add rustfmt
```
```suggestion
      - uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: stable
          override: true
          components: rustfmt
```

mbrobbel commented on .github/workflows/pr.yml:
```diff
@@ -64,3 +64,59 @@ jobs:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+      - run: rustup component add clippy
```
```suggestion
      - uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: stable
          override: true
          components: clippy
```

mbrobbel commented on validator/build.rs:
```diff
@@ -0,0 +1,19 @@
+extern crate prost_build;
+extern crate walkdir;
+
+use std::ffi::OsStr;
+
+fn main() {
+    let proto_path = "../proto";
+
+    let proto_files: Vec<_> = walkdir::WalkDir::new(proto_path)
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| {
+            e.path().extension() == Some(OsStr::new("proto")) && e.metadata().unwrap().is_file()
+        })
+        .map(|e| e.into_path())
+        .collect();
+
+    prost_build::compile_protos(&proto_files, &[proto_path]).unwrap();
+}
```
```suggestion
use std::{ffi::OsStr, io::Result};

fn main() -> Result<()> {
    let proto_path = "../proto";

    let proto_files: Vec<_> = walkdir::WalkDir::new(proto_path)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.path().extension() == Some(OsStr::new("proto")) && e.metadata().unwrap().is_file()
        })
        .map(|e| e.into_path())
        .collect();

    prost_build::compile_protos(&proto_files, &[proto_path])
}
```

jvanstraten committed:
Apply suggestions from code review

Co-authored-by: Matthijs Brobbel <m1brobbel@gmail.com>

mbrobbel commented on validator/build.rs:
```diff
@@ -0,0 +1,19 @@
+extern crate prost_build;
+extern crate walkdir;
+
```
```suggestion
```
You don't need these.

jvanstraten committed:
Remove section from pre-commit that doesn't work (yet)

jvanstraten committed:
Update validator/build.rs

Co-authored-by: Matthijs Brobbel <m1brobbel@gmail.com>

jvanstraten committed:
Create some initial versions of some helper types for the validator

jvanstraten committed:
WIP; add a bunch more helper types and some basic data type stuff

jvanstraten committed:
WIP

jvanstraten committed:
Add some test for derived protobuf metadata

jvanstraten committed:
WIP

jvanstraten committed:
WIP

jvanstraten committed:
WIP

jvanstraten committed:
WIP refactoring context

jvanstraten committed:
WIP

jvanstraten committed:
The first tree traversal is a fact

jvanstraten committed:
WIP, doesn't compile

jvanstraten committed:
WIP: working again

jvanstraten committed:
WIP: finish all anchor definition logic

jvanstraten committed:
Reorganize validation code

jvanstraten committed:
Initial version for HTML output

jvanstraten committed:
Add protobuf-based output format

jvanstraten committed:
Some cleanup

jvanstraten committed:
Add C interface

jvanstraten committed:
Updates to HTML output, and cleanup in general

jvanstraten committed:
Add python bindings, and more cleanup

jvanstraten committed:
Add CMake support for C bindings and add test

jvanstraten committed:
clang-format protobuf and validator C test file

jvanstraten committed:
Fix protobuf lint violations

jvanstraten committed:
Preparing to add YAML resolution and parsing logic

jvanstraten committed:
Add YAML loading capabilities

jvanstraten committed:
Downgrade (and pin) Corrosion version

jvanstraten committed:
Remove static linking option for validator C bindings

For now anyway, because the dep on curl breaks it. All the YAML
stuff should be put behind a feature flag in the future, so you
can still statically link with the feature disabled.

jvanstraten committed:
Remove files that shouldn't have been checked in

jvanstraten committed:
WIP improving Python API

jvanstraten committed:
Restructure, WIP on CI again

jvanstraten committed:
Test CI again

jvanstraten committed:
Remove reliance on Python protobuf-gen

jvanstraten committed:
Canonicalize more paths to try to get Windows CI to work

jvanstraten committed:
Test CI

jvanstraten committed:
Remove dependency on libcurl

jvanstraten committed:
Fix error message

jvanstraten committed:
Strip UNC prefixes in Windows paths passed to protoc

jvanstraten committed:
Use typing.Iterable for type hint

jvanstraten committed:
Extreme prejudice for pip commands on mac runner

jvanstraten committed:
Missed one of the canonicalize() calls

jvanstraten committed:
Increase pip/python executable resolution prejudice

jvanstraten committed:
Re-enable C static linking now that libcurl is out of the way

jvanstraten committed:
Restructure again

jvanstraten committed:
Make sure latest versions of python deps are available in CI

jvanstraten committed:
Fix clippies for new Rust release

jvanstraten committed:
Finish and document diagnostic handling

jvanstraten committed:
Implement YAML tree walking

jvanstraten committed:
Improve validator configurability in Rust

jvanstraten committed:
Expose configuration structures to C and Python

jvanstraten committed:
Fix C test

jvanstraten committed:
Refactor into a sane(r) module tree and document all modules

jvanstraten committed:
Fix Python tests

jvanstraten committed:
Support patterns for Config::whitelisted_any_urls

jvanstraten committed:
Remove validator lambdas in traversal code

jvanstraten committed:
Deduplicate some code in parse::traversal

jvanstraten committed:
Deduplicate more code in parse::traversal

jvanstraten committed:
Document parse module

jvanstraten committed:
Merge branch 'main' into validator

jvanstraten committed:
Add license data

mbrobbel commented on .github/workflows/validator-py.yml:
```diff
@@ -0,0 +1,94 @@
+      uses: actions/upload-artifact@v2
+      with:
+        name: wheels
+        path: dist
```
You can use a matrix here. [Reference](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-running-with-multiple-operating-systems).

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+          profile: minimal
+          toolchain: stable
+          override: true
+      - name: Run
```
```suggestion
      - name: Check
```

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+          toolchain: stable
+          override: true
+      - name: Run
+        run: cargo test
```
Same as above. Use matrix for this.

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+          profile: minimal
+          toolchain: stable
+          override: true
+      - name: Run
```
```suggestion
      - name: Test
```

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+          toolchain: stable
+          override: true
+          components: rustfmt
+      - name: Run
```
```suggestion
      - name: Rustfmt
```

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+          toolchain: stable
+          override: true
+          components: clippy
+      - name: Run
```
```suggestion
      - name: Clippy
```

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -0,0 +1,81 @@
+name: Validator (Rust)
```
Maybe add https://github.com/actions/cache/blob/main/examples.md#rust---cargo to these jobs?

jvanstraten replied:
Closing because there's a new discussion for this below.

mbrobbel commented on .github/workflows/validator-c.yml:
```diff
@@ -0,0 +1,23 @@
+name: Validator (C)
+
+on:
```
Maybe add path filters? https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-and-excluding-paths

jvanstraten replied:
I don't know if that's desirable... It's pretty easy to change code elsewhere and break something in the C bindings.

mbrobbel commented on validator/c/README.md:
```diff
@@ -0,0 +1,55 @@
+# C bindings for validator
+
+This directory contains a Rust/cbindgen project to generate C bindings for
+the validator library.
```
```suggestion
the validator crate.
```
🦀

mbrobbel commented on validator/c/README.md:
```diff
@@ -0,0 +1,55 @@
+
+## Installation
+
+No binaries are published yet at this time.
```
```suggestion
No binaries are published yet.
```

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -0,0 +1,609 @@
+// Functions dereferencing raw pointers are kind of par for the course in a C
+// interface.
+#![allow(clippy::not_unsafe_ptr_arg_deref)]
```
👎

jvanstraten replied:
It's the lesser evil. Marking the functions unsafe like clippy wants means that you can no longer selectively mark parts of the function bodies unsafe, which does much more for readability than a catch-all unsafe marker. It also doesn't at all matter whether these functions are marked unsafe or not, as they will only ever be called from C. Added comment explaining the reasoning.

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -0,0 +1,609 @@
+    true
+}
+
+/// Whitelists a protobuf message type for use in advanced extensions. If an
```
Maybe use allowlist and blocklist?

jvanstraten replied:
Right, forgot "whitelist" is a trigger word nowadays...

A blocklist doesn't make much sense in this context though, if that's what you meant. According to the Substrait spec, anything is allowed here, so the validator should never throw an error, and it already has to throw a warning by default because it can never guarantee correctness of an extension based on protobuf Any.

0182437

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -0,0 +1,609 @@
+        std::cmp::Ordering::Greater => substrait_validator_core::Level::Info,
+        std::cmp::Ordering::Equal => substrait_validator_core::Level::Warning,
+        std::cmp::Ordering::Less => substrait_validator_core::Level::Error,
+    }
```
I find something like this easier to understand/read: 
```suggestion
    match x {
        1..=i32::MAX => substrait_validator_core::Level::Info,
        0 => substrait_validator_core::Level::Warning,
        i32::MIN..=-1 => substrait_validator_core::Level::Error,
    }
```

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -0,0 +1,609 @@
+    data.push(0);
+
+    // Make sure vec.len() == vec.capacity() so we don't have to store both.
+    data.shrink_to_fit();
```
This does not give that guarantee: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.shrink_to_fit.  You should convert to a boxed slice: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice. And instead of the mem::forget below you should use https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html here.

jvanstraten replied:
Converting a boxed slice to a raw pointer gives a `*mut [u8]`, which can be cast to `*mut u8` but can't be cast back. I couldn't immediately find a from_raw_parts for it either (assuming the issue is that the length needs to be known to cast back). There is apparently a whole bunch of magic and special-casing involved with making boxed slices work in Rust (without special casing, the slice a.k.a. fat pointer would be something like `(*mut u8, usize)`, so a boxed slice would be a pointer to that thing, but that's not how it works). Ultimately, I'm not at all confident that I can get this working without accidentally relying on some kind of UB.

Instead I just rewrote the code to also store the vector capacity, so the vector can be recovered "safely". I also added comments explicitly stating what assumptions each `unsafe {}` block makes based on our slack discussion.

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -0,0 +1,609 @@
+
+    // Recover the vector and drop it.
+    unsafe {
+        drop(Vec::from_raw_parts(data, len, len));
```
See comment above about capacity.

jvanstraten committed:
WIP on Python distribution issues

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        !*self
```
```suggestion
        *self == false
```

jvanstraten replied:
I actually have to revert this because clippy doesn't agree with you.

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
```
Shouldn't this be:
```suggestion
        *self == 0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
```
```suggestion
        *self == 0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
```
```suggestion
        *self == 0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
```
```suggestion
        *self == 0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0.0
```
```suggestion
        *self == 0.0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0.0
```
```suggestion
        *self == 0.0
```

mbrobbel commented on validator/core/src/input/proto.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Bytes(vec![])
```
You could also use the Default trait for all these impls here.

jvanstraten replied:
Sure, I guess, but Rust only "incidentally" has the same default value definitions that protobuf has, so I prefer explicitly stating what protobuf considers to be the default value for a type. Unless I'm misunderstanding what you mean?

mbrobbel commented on validator/core/src/input/traits.rs:
```diff
@@ -0,0 +1,269 @@
+
+    /// Returns the name of the selected variant of a oneof field, if this
+    /// is a rust enum used to represent a oneof field.
+    fn oneof_variant(&self) -> Option<&'static str>;
```
Maybe you can use an associated constant for this?

jvanstraten replied:
Uh, I'm not sure how, the value depends on `&self`.

mbrobbel commented on validator/core/src/lib.rs:
```diff
@@ -0,0 +1,117 @@
+pub mod input;
+
+// Aliases for common types used on the crate interface.
+pub type Config = input::config::Config;
```
You can also pub use them:
```
mod input;
pub use input::config::Config;
...
```

jvanstraten replied:
derp, that's much nicer. The C++ programmer in me didn't even consider trying to re-export something that isn't a "namespace."

mbrobbel commented on validator/core/src/lib.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+/// Iterates over all diagnostic messages in the tree.
+pub fn iter_diagnostics(root: &Node) -> impl Iterator<Item = &Diagnostic> + '_ {
```
Why not make this and the other methods members of Node?

jvanstraten replied:
I'm kind of on the fence about this. I like the idea of putting all the functions you would need as a normal user to get basic validation going in the toplevel module of the crate, so it's easy to find where the entry points for the library are. But maybe, like `iter_diagnostic()`, they should all *also* be methods of `Node`.

I could also make the root of the tree an explicit type at the top level of the crate and implement these methods on it, since these methods are only really intended to be used for a tree as a whole, rather than on any subtree. But that doesn't feel right either, since it's not necessarily wrong to apply them to subtrees, either.

Maybe both? It hasn't been feeling right to pass `Node` around as the complete representation of a tree. In the C and Python bindings I needed wrappers anyway, where I called the wrapper `substrait_validator_result_handle` and `ParseResult` respectively. Might as well do the same in Rust, even if it would only be a thin wrapper there (unless at some point additional root-only info needs to be passed, so it might be a good idea to make a struct for that already for forward-compatibility's sake, especially in the proto serialization of this).

Guess I have to mess around with this a bit more.

mbrobbel commented on validator/core/src/output/data_type.rs:
```diff
@@ -0,0 +1,356 @@
+        if let Some(variation) = &self.variation {
+            write!(f, "[{}]", variation)?;
+        }
+        if self.parameters.is_empty() {
```
I guess this should be:
```suggestion
        if !self.parameters.is_empty() {
```

mbrobbel commented on validator/core/src/output/data_type.rs:
```diff
@@ -0,0 +1,356 @@
+
+impl std::fmt::Display for Class {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Class::Simple(simple) => write!(f, "{}", simple),
+            Class::Compound(compound) => write!(f, "{}", compound),
+            Class::UserDefined(user_defined) => write!(f, "{}", user_defined),
+            Class::Unresolved(name) => write!(f, "{}!", name),
+        }
```
```suggestion
       write!(f, "{}", match self {
            Class::Simple(simple) => simple,
            Class::Compound(compound) => compound,
            Class::UserDefined(user_defined) => user_defined,
            Class::Unresolved(name) => name,
        })
```

jvanstraten replied:
Doesn't work because simple, compound, user_defined, and name are all different types.

mbrobbel commented on validator/core/src/output/data_type.rs:
```diff
@@ -0,0 +1,356 @@
+                    ));
+                }
+                if let Parameter::Unsigned(length) = params[0] {
+                    if length < 1 && length > 2147483647 {
```
```suggestion
                    if length < 1 && length > i32::MAX {
```

jvanstraten replied:
I think I did it this way because the spec explicitly lists this as the range. It might as well have listed 10000000 or whatever.

mbrobbel replied:
We can put one of them in a comment? One matches the spec, the other makes it easier to understand where that number comes from.

jvanstraten replied:
Added comment and used a local const for it to deduplicate.

mbrobbel commented on validator/core/src/output/data_type.rs:
```diff
@@ -0,0 +1,356 @@
+                    if length < 1 && length > 2147483647 {
+                        return Err(cause!(
+                            TypeMismatchedParameters,
+                            "{} length {} is out of range 1..2147483647",
```
https://doc.rust-lang.org/std/primitive.i32.html#associatedconstant.MAX

jvanstraten replied:
Same here.

mbrobbel commented on validator/core/src/output/path.rs:
```diff
@@ -0,0 +1,202 @@
+                primitive_data::as_ident_or_string(field),
+                primitive_data::as_ident_or_string(variant)
+            ),
+            PathElement::Index(index) => write!(f, "[{}]", index),
```
Random observation that probably applies to other format strings, but you can also do this now:
```suggestion
            PathElement::Index(index) => write!(f, "[{index}]"),
```

jvanstraten replied:
Cool, didn't know that was a thing.

jvanstraten committed:
Add missing license header

jvanstraten committed:
Fix missing dependency in python workflows

jvanstraten committed:
Fix path issue

jvanstraten committed:
Apply suggestions from code review

Co-authored-by: Matthijs Brobbel <m1brobbel@gmail.com>

jvanstraten committed:
Fix sdist workflow

jvanstraten committed:
Fix clippy

jvanstraten committed:
Revise terminology

jvanstraten committed:
Fix sdist

jvanstraten committed:
Fix UB in C interface layer and more thoroughly document unsafe {} blocks

mbrobbel commented on validator/c/src/lib.rs:
```diff
@@ -490,53 +530,97 @@ fn export(
+    // pointer, length, and capacity do indeed form the raw parts of a valid
+    // Vec.
+    let vec = unsafe { Vec::from_raw_parts(buffer_ptr, length, capacity) };
+    drop(vec);
```
You don't need this `drop` but I like it.

jvanstraten replied:
Haha, yeah, more explicit more better, right? It was there before too, I just pulled it out of the `unsafe {}`.

jvanstraten committed:
Use matrices to reduce repetition in workflow files

jvanstraten committed:
Re-export instead of typedef

jvanstraten committed:
Use string interpolation syntax to make format! strings easier to read

jvanstraten committed:
String interpolation and extra comments for output::data_type

jvanstraten committed:
Use CARGO_MANIFEST_DIR env var in build.rs and add rerun-if-changed commands

jvanstraten committed:
Add dep versions for cargo package

jvanstraten committed:
Disable unneeded features of jsonschema crate

jvanstraten committed:
Revise Rust packaging

jvanstraten committed:
Add ParseResult abstraction to Rust

jvanstraten committed:
Add ParseResult abstraction to protobuf serialization of validation results

jvanstraten committed:
Minor API function refactoring

jvanstraten committed:
Add rust cache and cargo doc check to CI

jvanstraten committed:
Fix CI?

mbrobbel commented on .github/workflows/validator-rs.yml:
```diff
@@ -73,7 +73,6 @@ jobs:
           toolchain: stable
           override: true
-          components: rustdoc
       - uses: Swatinem/rust-cache@v1
```
Why don't you use the official cache action: https://github.com/actions/cache/blob/main/examples.md#rust---cargo?

jvanstraten replied:
Because (in no particular order):

 - it's a lot shorter/less cluttered/less magic-stringy than calling the cache action directly;
 - it doesn't require me to think about whether those cache keys and magic paths are correct because the defaults of this action should already be;
 - it's actually documented in the context of Cargo/Rust beyond just throwing an example out there and calling it a day;
 - it was the top hit on google;
 - I'm pretty sure I've seen it used somewhere before, though I'm not sure where.

Is there anything wrong with it?

mbrobbel replied:
No, not necessarily anything wrong with it. But I'd like to suggest using offical actions (when available/possible) because it is more likely that they are being maintained.

jvanstraten replied:
I mean... this action is just a thin wrapper around the official method, I'm pretty sure. So even if it's never updated, it's not worse than copypasting the (current) official method into the workflow directly. On the other hand, if ever something changes with cargo that requires changes to which directories need to be cached, there is at least a chance that the this third party action will be updated.

jvanstraten committed:
Fix Python tests

jvanstraten committed:
Initial commit for validator CLI in Python

jvanstraten committed:
Add Python dependency to CI workflow

jvanstraten committed:
Add validator config support to CLI

jvanstraten committed:
Replace pycurl with urllib, because pycurl doesn't ship self-contained wheels

jvanstraten committed:
Add additional Python API tests

jvanstraten committed:
Add CLI test cases

jvanstraten committed:
Add missing license headers

jvanstraten committed:
Fix tests not using installed library

jvanstraten committed:
Fix file:// URL in test on Windows

jvanstraten committed:
Bump JSOM version to pull in bugfix, enable roundtrip tests

jvanstraten committed:
WIP on type & relation validation

jvanstraten committed:
Fix cargo doc warnings

jvanstraten committed:
Add schema tracking logic

jvanstraten committed:
Factor lots of stuff out of traversal macros and into Context

jvanstraten committed:
Wrap DataType in Arc everywhere, because they were getting cloned a lot

jvanstraten committed:
Refactor extension parse & resolution logic

jvanstraten committed:
Extend documentation features for nodes

jvanstraten committed:
Add additional descriptive information

jvanstraten committed:
Minor: only emit data type nodes when the data type is actually known

jvanstraten committed:
Add validation test runner

jvanstraten committed:
Use protoc from prost_build for test runner

jvanstraten committed:
Force flushing stdout when printing in test runner

jvanstraten committed:
Add pyyaml to test runner dependencies

jvanstraten committed:
Clean up comments

jvanstraten committed:
WIP on expressions

jvanstraten committed:
Implement the most important bits of expression validation

jvanstraten committed:
Finish validation for read relation

jvanstraten committed:
Fix URI regexes, fix proto name derivation insofar possible

jvanstraten committed:
Improve glob syntax regex, downgrade URI errors to warnings

jvanstraten committed:
Implement sorts and basic window function validation

jvanstraten committed:
Implement validation for all relations except aggregations

jvanstraten committed:
Minor fixes to validation, embed FontAwesome according to license

jvanstraten committed:
Fix formatting of comment blocks in HTML

jvanstraten committed:
Add shorter description for basic fetch relation

jvanstraten committed:
Minor fixes to validation and HTML output

jvanstraten committed:
Match names in RelRoot with schema

jvanstraten committed:
Fix grammar in output

jvanstraten committed:
Implement validation for subqueries

jvanstraten committed:
Use crate for URI parsing, clean up long string literals

jvanstraten committed:
Merge remote-tracking branch 'upstream/main' into validator

jvanstraten committed:
Resolve FIXMEs related to URIs, as docs have been added upstream since

jvanstraten committed:
Fix divisions by zero in string_util

jvanstraten committed:
Disable field name matching error when incoming type is (partially) unresolved

jvanstraten committed:
Implement validation for aggregate relation

jvanstraten committed:
Emit data types recursively for HTML export format

jvanstraten committed:
Improve documentation somewhat

jvanstraten committed:
Add APIs for retrieving the list of defined diagnostic codes

jvanstraten committed:
Add type variation base class information, remove unresolved type description (wasn't using it)

jvanstraten committed:
Various improvements for plumbing related to parsing

jvanstraten committed:
Update test case (really need to make better ones already)

jvanstraten committed:
Make cargo package work

jvanstraten committed:
Document and clean up dependencies

jvanstraten committed:
Document release procedure, some cleanup

jvanstraten committed:
Update release procedure

jvanstraten committed:
Improve/fix Python API test case

jvanstraten committed:
Update JSOM (decoder is broken, but encoder is much nicer now)

jvanstraten committed:
Fix sdist release logic

jvanstraten committed:
Allow different branches of if/else statements to differ in nullability

jvanstraten committed:
Add some features to the test runner and add positive test cases for some TPC-H queries

jvanstraten committed:
Build test runner with optimizations and disable HTML gen in workflow

jvanstraten committed:
Break long YAML strings

jvanstraten committed:
Add yamllint to pre-commit hooks

jvanstraten committed:
Improve performance for validation tests once more

jvanstraten committed:
Add validation tests for all literals and types, fix many bugs

jvanstraten committed:
Add test cases for scalar references

jvanstraten committed:
Implement "incremental compilation" for Python part of test runner

jvanstraten committed:
Implement half-decent CLI for test runner

jvanstraten committed:
Clean up Rust code for test runner, implement parallel testing

jvanstraten committed:
Upgrade JSOM, re-enable tests

jvanstraten committed:
Add more complex plan to Python test cases

jvanstraten committed:
Add tests for masked and outer references

jvanstraten committed:
Add tests for conditional expressions

jvanstraten committed:
Add test cases for subqueries and reorganize directory structure

jvanstraten committed:
Fix duplicate grouping sets in TPC-H query tests

jvanstraten committed:
Some more bugfixes to Isthmus-generated plans, implement TPC-H 2, 4 & 7 manually

jvanstraten committed:
Merge remote-tracking branch 'upstream/main' into validator

jvanstraten committed:
Make ifElse return nullable when any condition is a nullable bool

jvanstraten committed:
decimal vs dec in function signatures

jvanstraten committed:
Reorder extensionUris in TPC-H queries (no functional change)

jvanstraten committed:
Manually implement TPC-H8

jvanstraten committed:
Manually implement TPC-H9

jvanstraten committed:
Add black and flake8 for Python code

jvanstraten committed:
Add some test cases for ReadRel, improve specificity of some diag codes

jvanstraten committed:
Add test cases for ReadRel, RelCommon, and PlanRel

jvanstraten committed:
Add test cases for filter and sort relations

jvanstraten committed:
Add test cases for project and cross relations

jvanstraten committed:
Add test cases for join relations

jvanstraten committed:
Fix test case names

jvanstraten committed:
Add test cases for set and fetch relations

jvanstraten committed:
Add test cases for aggregates and advanced extensions

jvanstraten committed:
Merge remote-tracking branch 'upstream/main' into validator

jvanstraten committed:
Merge remote-tracking branch 'upstream/main' into validator

jvanstraten committed:
Downgrade Clippy to 1.59 in CI, since 1.60 throws an ICE

jvanstraten committed:
Lock jsonschema to 0.15.0, see https://github.com/Stranger6667/jsonschema-rs/pull/357

jvanstraten committed:
Remove superfluous error message for missing nullability information in types

jvanstraten committed:
Fix grammar in summary of sort relation

jvanstraten committed:
rm -rf to point people to the correct location
